Fix compiler optimizing decrypt xorstring function

TotallyNotElite opened the pull request:
nullworks/cathook#487 got me interested in using xorstrings. After implementing them in a test project and reverse engineering it, I discovered that the string was still visible (most likely due to compile time optimization). Adding a volatile statement to the loop fixes this problem. (NOTE: I did not test these changes in fuzion (only in nullworks/cathook))

LWSS commented:
It does work, according to IDA strings window. I remember adding the __always_inline specifier to it.

I was actually thinking of removing XORSTR because it makes compiling slower, and not all the strings are even xorstr'd lol, I'm not sure it does anything.

Now that i've got this open in IDA though, i'm reminded that the function names are in the binary, did anyone figure out how to get rid of those?

FilipeSilvens commented:
yes @LWSS i did this in my cheat ages ago.
add `-s and -fvisibility=hidden` to the gcc flags in cmakelists

LWSS commented:
@FilipeSilvens 
damn ur better than Valve at hiding those, thanks

TotallyNotElite commented:
@LWSS xorstrings work fine for me? 
![image](https://user-images.githubusercontent.com/38938720/50575795-13b89500-0e05-11e9-8aa1-dcc9eb82a267.png)

TotallyNotElite commented:
I've added some changes.

- CMake: -ggdb instead of -g (this allows gdb to get line numbers when running bt)
- CMake: -shared is unnecessary since add_library is used below
- CMake: -fvisibility-inlines-hidden stop even more symbols from being exported
- CMake: CMAKE_CXX_FLAGS instead of copy pasting between ..._RELEASE and ..._DEBUG
- Readme: Removed dependency "v8", it was never used anywhere? What was the point?

LWSS commented:
v8 is for stuff like this, (https://github.com/LWSS/Fuzion/blob/master/src/SDK/panorama/PanoramaUIEngine.h)

So I could fully use the v8 types from the library for the prototypes.

But I guess I didn't actually do that, so we'll just leave it on the table.

I like the cmake changes so i'll merge

BenCat07 commented:
@LWSS These were most likely fucked by gcc optimization, bu the XorStrings are just a tad annoying, i found out myself that this one stores the XORKEY you can use with
`bytes[t]^ = XORKEY+t` to decrypt the bytes visibly in many cases, e.g. a 
```do
{
    v32[t] ^= t + 51;
}
while (t =< 0x13)```
So it may not even be that useful even when used properly

BenCat07 commented:
Oh nvm, in your case you are fine as you don't try to obfuscate code from humans but machines

TotallyNotElite commented:
@bencat07 A reverse engineer can reverse xorstrings, but its basically impossible for vac to do the same.
